dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +199,303 @@ cdef int _local_tracker(
+          DirectionGetter dg,
+          ConstrainedTissueClassifier tc,
+          double* voxel_size,
+          double step_size,
+          TissueClass * tissue_class,
+          int pft_nbr_steps,
+          int particle_count,
+          np.float_t[:, :, :, :] particle_paths,
+          np.float_t[:, :, :, :] particle_dirs,
+          np.float_t[:] particle_weights,
+          np.int_t[:, :, :] particle_states):
+    cdef:
+        double sum_weights, sum_squared, N_effective, rdm_sample
+        double point[3]
+        double dir[3]
+        double voxdir[3]
+        double eps = np.finfo(np.float).eps
+        int s, p, j
+
+    if pft_nbr_steps <= 0:
+        return streamline_i
+
+    for p in range(particle_count):
+        copypoint(&streamline[streamline_i, 0], &particle_paths[0, p, 0, 0])
+        copypoint(&directions[streamline_i, 0], &particle_dirs[0, p, 0, 0])
+        particle_weights[p] = 1. / particle_count
+        particle_states[0, p, 0] = TRACKPOINT
+        particle_states[0, p, 1] = 0
+
+    for s in range(pft_nbr_steps):
+        for p in range(particle_count):
+            if particle_states[0, p, 0] != TRACKPOINT:
+                for j in range(3):
+                    particle_paths[0, p, s, j] = 0
+                    particle_dirs[0, p, s, j] = 0
+                continue  # move to the next particle
+            copypoint(&particle_paths[0, p, s, 0], point)
+            copypoint(&particle_dirs[0, p, s, 0], dir)
+
+            if dg.get_direction_c(point, dir):
+                particle_states[0, p, 0] = INVALIDPOINT
+                particle_weights[p] = 0
+            else:
+                for j in range(3):
+                    voxdir[j] = dir[j] / voxel_size[j]
+                fixed_step(point, voxdir, step_size)
+                copypoint(point, &particle_paths[0, p, s + 1, 0])
+                copypoint(dir, &particle_dirs[0, p, s + 1, 0])
+                particle_states[0, p, 0] = tc.check_point_c(point)
+                particle_states[0, p, 1] = s + 1
+                particle_weights[p] *= 1 - tc.get_exclude_c(point)
+                if particle_weights[p] < eps:
+                    particle_weights[p] = 0
+                if (particle_states[0, p, 0] == INVALIDPOINT and
+                        particle_weights[p] > 0):
+                    particle_states[0, p, 0] = TRACKPOINT
+
+        sum_weights = 0
+        for p in range(particle_count):
+            sum_weights += particle_weights[p]
+        sum_squared = 0
+        for p in range(particle_count):
+            particle_weights[p] = particle_weights[p] / sum_weights
+            sum_squared += particle_weights[p] * particle_weights[p]
+
+        # Resample the particles if the weights are too uneven.
+        # Particles with negligable weights are replaced by duplicates of
+        # those with high weigths through resampling
+        N_effective = 1. / sum_squared
+        if N_effective < particle_count / 10.:
```

Comment:
**Using your test** (without the fixed random), I couldn't get this to happen. In fact, everytime I tested, all elements in the arrays (paths, dirs, states, weights) are perfectly equal in the `p` dimension. Looking at the code, I don't see how they could be different. If they are actually always equal, then this algo is much too complicated!

I probably don't understand well enough. Can you please explain how it can happen? Is it only because I'm using test data?

Reply:
On my side, the resampling happens and the weights, dirs, states, path are different.
try running `nosetests --nocapture -v` in `dipy/tracking/local/`
and adding a print at line localtrack.pyx:463 `print("Resampling")`
add adding a print at line localtracking.py:239 `print(self.particle_particle_weights)`

You can also try running the example: `python doc/example/particle_filtering_tractography.py` to check if `pft_streamlines` and `'local_prob_streamlines` are different.


----------------

In `_pft(.)`, we do probabilistic tractography for P particles and N steps. at each tracking step, the P particle weights are updates following the information provided by the `TissueClassifier` (localtrack.pyx:line 441). After the N steps, we sample 1 particle from the P particles using their weight (higher weight, more likely to be sampled). The path of this particle is the one used in the final streamline.

The thing here is that P particles might be very few particles to properly explore a given configuration. To overcome that, we focus the _exploration_ in regions more promising (i.e. particles with higher weights). To do that, we discard particles with low weight and duplicate particles with high weights. This is done when the weight distribution is unbalanced (e.g. when `1. / sum_squared < particle_count / 10`).